src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let newState = state;
+
+  for (const action of actions) {
+    newState = { ...newState };
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = addProperties(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        newState = removeProperties(newState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
```

Comment:
[CODE STYLE]: switch/case should always have default case for error handling. Add a default case to handle any unexpected action types.